toolgenie.py

```diff
@@ -0,0 +1,219 @@
+#!/usr/bin/env python
+from __future__ import print_function
+from collections import namedtuple
+import os, re, readline
+import xml.etree.ElementTree as ET
+
+# https://stackoverflow.com/questions/21731043/use-of-input-raw-input-in-python-2-and-3
+# Bind raw_input to input in Python 2.
+try:
+    input = raw_input
+except NameError:
+    pass
+
+class CMSSW(namedtuple('CMSSW', 'type major mid minor extra note')):
+    __slots__ = ()
+    def __eq__(self,type,major,mid,minor,extra,note):
+        return self.type == type and self.major == major and self.mid == mid and self.minor == minor and self.extra == extra and self.note == note
+CMSSW.__new__.__defaults__ = ("",) * len(CMSSW._fields)
+
+class Release(namedtuple('Release', 'architecture label type state prodarch')):
+    __slots__ = ()
+    def __eq__(self, architecture, label):
+        return self.architecture == architecture and self.label == label
+    def getReleaseString(self):
+        return "architecture=%s;label=%s;type=%s;state=%s;prodarch=%s;" % (self.architecture,self.label,self.type,self.state,self.prodarch)
+
+class Toolbox(namedtuple('Toolbox',['Release','Tools','Path'])):
+    __slots__ = ()
+
+class Tool(namedtuple('Tool',['Name','ConfigPaths','Locations','Versions','Architectures','Releases'])):
+    __slots__ = ()
+
+def filter_on_architecture(release_map, architectures):
+    return [release for release in release_map if any(architecture in release.architecture for architecture in architectures)]
+
+def filter_on_label(release_map, labels):
+    return [release for release in release_map if any(label in release.label for label in labels)]
+
+def filter_on_tool(toolboxes, tool):
+    return [tb for tb in toolboxes if tool in tb.Tools]
+
+def get_architectures(release_map):
+    archs = []
+    for rel in release_map:
+        if rel.architecture not in archs:
+            archs.append(rel.architecture)
+    return sorted(archs)
+
+def get_labels(release_map):
+    labels = []
+    for rel in release_map:
+        if rel.label not in labels:
+            labels.append(rel.label)
+    return sorted(labels,key=lambda label: CMSSW([int(x) if x.isdigit() else 999 if type(x)==str else x for ix,x in enumerate(label.split("_"))]))
+
+def get_tools(toolbox_list):
+    tools = []
+    for toolbox in toolbox_list:
+        for tool in toolbox.Tools:
+            if tool not in tools:
+                tools.append(tool)
+    return sorted(tools)
+
+def parse_release_map():
+    relmap = []
+    with open("/cvmfs/cms.cern.ch/releases.map",'r') as release_map:
+        for line in release_map:
+            line_list = process_map_line(line)
+            # Skip releases listed in the map file who are names after branches. These tend to not have actual folders on CVMFS.
```

Comment:
maybe point out explicitly in this comment that you're targeting names in the format of CMSSW_#_#_X

Reply:
Done